### PR TITLE
Add `GroupByMetricSpec`, `GroupByMetricInstance`, and `LinkableMetric`

### PR DESCRIPTION
**Pull request by courtneyholcomb**
### Description
Adding some functionality that will be used for metrics as dimensions. Note that this only allows metrics to be joined via entity links (the same way other linkable elements are joined). Later we will need to enable joining metrics via dimensions.
This is only step one - coming next:
- instance converters
- source node selection for metric group by
- end to end query execution for group by metrics in filters
- end to end query execution for group by metrics in group by

Skipping changelog until the feature is complete.

### CODE REVIEW

**Comment by courtneyholcomb** [comment]
Looks like I need to update my `noqa` comments here (hence failing linter) but that shouldn't block review!

**Comment by tlento** [comment]
> Looks like I need to update my noqa comments here (hence failing linter) but that shouldn't block review!

🤦 sorry about that.

**Review by tlento — COMMENTED** [review]
Sorry I didn't really get to go through this and I have to leave now. :( Will pick it up first thing Monday.

File: metricflow/test/fixtures/dbt_projects/metricflow_testing/logs/dbt.log.1 [hunk 1]
```diff
@@ -0,0 +1,1387 @@
+[0m16:29:47.488269 [debug] [MainThread]: Re-using an available connection from the pool (formerly MetricFlow_request_mf_rid__7ecsbtuo, now MetricFlow_request_mf_rid__wrlve1tb)
```
[the remark this answers is not in the document]
Comment (courtneyholcomb) [reply metricflow/test/fixtures/dbt_projects/metricflow_testing/logs/dbt.log.1]: Shoot, I got rid of this once already but maybe a rebase put it back 🤔 will remove again!

[the remark this answers is not in the document]
Comment (courtneyholcomb) [reply metricflow/test/fixtures/dbt_projects/metricflow_testing/logs/dbt.log.1]: Oh nevermind - your comment was on an outdated version, this is gone now!

**Review by tlento — APPROVED** [review]
So far this looks reasonable to me, looking forward to the rest of this stack of changes!

File: metricflow/model/semantics/linkable_spec_resolver.py [hunk 2]
```diff
@@ -171,9 +209,17 @@ def intersection_by_path_key(linkable_element_sets: Sequence[LinkableElementSet]
             ]
         )
 
+        common_linkable_metric_path_keys: Set[ElementPathKey] = set.intersection(
+            *[
```
[the remark this answers is not in the document]
Comment (courtneyholcomb) [reply metricflow/model/semantics/linkable_spec_resolver.py:213]: Interesting! I just did some local testing and haven't been able to reproduce that error. Both `set().intersection(*[])` and `set().intersection()` just return an empty set. That being said, I'm specifically testing Pythono 3.9, so who knows about other versions. I'll update this loop and the others above to use a different syntax.

**Pull request merged**